pkg/storage/cacher.go

```diff
@@ -417,18 +401,13 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 
 // Returns resource version to which the underlying cache is synced.
 func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
-	// To avoid situation when LastSyncResourceVersion is processed before the
-	// underlying watchCache is propagated, we acquire 'usable' lock.
-	c.usable.RLock()
-	defer c.usable.RUnlock()
-
-	c.RLock()
-	defer c.RUnlock()
+	c.ready.wait()
```

Comment:
That's true, I think I added a test last time I fixed a bug here.

On Mon, Jun 6, 2016 at 4:33 PM, Xiang Li <notifications@github.com> wrote:

> In pkg/storage/cacher.go
> <https://github.com/kubernetes/kubernetes/pull/26860#discussion_r65988030>
> :
>
> > @@ -417,18 +401,13 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
> >
> >  // Returns resource version to which the underlying cache is synced.
> >  func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
> > -	// To avoid situation when LastSyncResourceVersion is processed before the
> > -	// underlying watchCache is propagated, we acquire 'usable' lock.
> > -	c.usable.RLock()
> > -	defer c.usable.RUnlock()
> > -
> > -	c.RLock()
> > -	defer c.RUnlock()
> > +	c.ready.wait()
>
> I am not sure why exactly we need to hold the c.Rlock. It would be great
> if @wojtek-t <https://github.com/wojtek-t> can help to review. Also if it
> is really tricky, I would love to write a test to ensure we make the tricky
> part right.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26860/files/5d6b423585362dd13620036bc44e1f3a1baec730#r65988030>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglox9DOYIzeq-y37zBQ5SN5_V8xwsks5qJK5NgaJpZM4IuX_4>
> .
>